File: introHPC/10/bern_worker.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <gmpxx.h>
#include <tuple>
#include <stack>
#include <utility>
#include <deque>
#include <chrono>
#include <thread>
#include <zmqpp/zmqpp.hpp>
#include <sstream>

using namespace std;

auto norm(mpz_class b, mpz_class p) -> mpz_class
{
    while (true) {
        if (b < 0) b += p;
        else if (b > p) b -= p;
        else break;
    }
    return b;
}

auto computePowerMod(const mpz_class &a, const long &k, const mpz_class &m) -> mpz_class
{
    int index = log2(k);
    mpz_class powmod = 1, powa = a, pow2 = 1;

    for (int i=0; i<=index; i++) {
        if ( (k & pow2) != 0) powmod *= powa;
        powa = (powa * powa) % m;
        pow2 <<= 1;
    }
    return norm(powmod % m, m);
}

auto computeInvertMod(const mpz_class &n, mpz_class p) -> mpz_class
{
    stack<tuple<mpz_class, mpz_class, mpz_class>> v;
    mpz_class a = p;
    mpz_class b = n;
    if (n > p) {
        a = n;
        b = p;
    }
    mpz_class d = a / b;
    mpz_class r = a % b;

    while (r > 1) {
        v.push(make_tuple(a, b, -d));
        a = b;
        b = r;
        d = a / b;
        r = a % b;
    }
    mpz_class m1 = 1, m2 = -d;
    mpz_class aa = a, bb = b;

    while (!v.empty()) {
        auto t = v.top();
        v.pop();
        a = get<0>(t);
        b = get<1>(t);
        d = get<2>(t);
        m1 += (m2 * d);
        bb = a;
        swap(m1, m2);
        swap(aa, bb);
    }
    mpz_class rr = (n == aa)? m1 : m2;
    return norm(rr, p);
}

auto computeOrderEp(const long &a, const long &p) -> long
{
    for (long k=1; k<p; k++) {
        if (computePowerMod(a, k, p) == 1) return k;
    }
    throw "computeOrderEp error";
}

auto computePrimitiveRoot(const long &p) -> long
{
    if (p == 2) return 1;
    for (int g=2; g<p; g++) {
        if (computeOrderEp(g, p) == (p - 1)) return g;
    }
    throw "computePrimitiveRoot error";
}

auto computeBkModP(const long &p, const long &k) -> mpz_class
{
    if (k == 0) return 1;
    if (k == 1) return (p == 2)? -1 : (p-1)/2;
    if (k &  1) return 0;
    if (p <= 3) return -1;

    long m = k % (p-1);
    if (m == 0) return -1;

    mpz_class g = computePrimitiveRoot(p);
    mpz_class r = computePowerMod(g, k-1, p);
    mpz_class u = (g + ((g % 2 == 1)? 0 : p) - 1) / 2;
    mpz_class S = 0, X = 1, Y = r;


    mpz_class hp = (p-1)/2, q, h, t;
    for (long i=1; i<=hp; i++) {
        q = norm( (g*X)/p, p);
        h = norm( (q-u)%p, p);
        t = norm( (h*Y)%p, p);
        S = norm( (S-t)%p, p);
        X = norm( (g*X)%p, p);
        Y = norm( (r*Y)%p, p);
    }
    mpz_class ans = computePowerMod(g, k, p);

    ans = computeInvertMod(p+1-ans, p);

    ans = norm( (S * ans) % p, p);
    ans = norm( (2 * ans) % p, p);
    ans = norm( (k * ans) % p, p);

    return ans;
}

using crt_t = tuple<mpz_class, long>;
using primeList_t = deque<long>;

auto worker(int threadNo) -> void
{
    zmqpp::context context;
    zmqpp::socket receiver(context, zmqpp::socket_type::pull);
    receiver.connect("tcp://localhost:6666");

    zmqpp::socket sender(context, zmqpp::socket_type::push);
    sender.connect("tcp://localhost:7777");

    cout << "worker " << threadNo << " ready." << endl;

    while (1) {
        string str;
        receiver.receive(str);
        stringstream ssIn;
        ssIn.str(str);
        long k, p;
        char slash;
        ssIn >> k >> slash >> p;
        mpz_class bkModP = computeBkModP(p, k);
        stringstream ssOut;
        ssOut << bkModP << "/" << p;
        sender.send(ssOut.str());
    }
}

auto main(int argc, char *argv[]) -> int
{
    auto nthreads = thread::hardware_concurrency();
    thread threads[nthreads];

    for (int i=0; i<nthreads; i++) {
        threads[i] = thread(worker, i);
    }
    for (int i=0; i<nthreads; i++) {
        threads[i].join();
    }
    return 0;
}

File: introHPC/10/bern_worker.py
from math import log

def computePowerMod(a, k, m): 
    index = int(log(k, 2)) 
    powmod = 1 
    powa = a
    pow2 = 1

    for i in range(0, index+1):
        if k & pow2 != 0:
            powmod *= powa
        powa = (powa * powa) % m
        pow2 = pow2 * 2
    return norm(powmod % m, m)
    
def computeOrderEp(a, p):
    for k in range(1, p):
        if computePowerMod(a, k, p) == 1:
            return k
    raise 
    
def computePrimitiveRoot(p):
    if p == 2:
        return 1
    for g in range(2, p):
        if computeOrderEp(g, p) == p - 1:
            return g
    raise
    
def norm(b, p):
    while (True):
        if (b < 0):
            b += p
        elif (b > p):
            b -= p
        else:
            break
    return b

def computeInvertMod(n, p):
    (a, b) = (n, p) if n > p else (p, n)
    d = a // b
    r = a % b
    v = []
    
    while r > 1:
        v = [(a, b, -d)] + v
        (a, b) = (b, r)
        d = a // b
        r = a % b
    
    (m1, aa, m2, bb) = (1, a, -d, b)
  
    for (a, b, d) in v:
        m1 += (m2 * d)
        bb = a
        (m1, aa, m2, bb) = (m2, bb, m1, aa)
    rr = m1 if n == aa else m2
    return norm(rr, p)     
                
def computeBkModP(p, k):
    if k == 0:
        return 1
    if k == 1:
        return -1 if p == 2 else (p-1)/2
    if (k & 1):
        return 0
    if (p <= 3):
        return -1
    m = k % (p-1)
    if m == 0:
        return -1
    g = computePrimitiveRoot(p)
    r = computePowerMod(g, k-1, p);
    u = (g + (0 if g & 1 else p) - 1) // 2
    S = 0 ; X = 1 ; Y = r
    for i in range(1, (p-1)//2 + 1):
        q = norm( (g * X) // p, p)
        h = norm( (q-u) % p, p)
        t = norm( (h * Y) % p, p)
        S = norm( (S - t) % p, p)
    
        X = norm( (g * X) % p, p)
        Y = norm( (r * Y) % p, p)
    ans = computePowerMod(g, k, p)
    ans = computeInvertMod(p + 1 - ans, p)
    ans = norm( (S * ans) % p, p)
    ans = norm( (2 * ans) % p, p)
    ans = norm( (k * ans) % p, p)
    return ans
    
def worker(threadNo):
    import zmq
       
    context = zmq.Context()
    receiver = context.socket(zmq.PULL)
    receiver.connect("tcp://localhost:6666")
    sender = context.socket(zmq.PUSH)
    sender.connect("tcp://localhost:7777")
    
    print("worker %s ready." % threadNo)
    while True:
        s = receiver.recv_string()
        list1 = s.split('/')
        k   = int(list1[0])
        pp  = int(list1[1])
        sender.send_string("%s/%s" % (computeBkModP(pp, k), pp))
   
if __name__ == "__main__":
    import multiprocessing as mp
    lock1 = mp.Lock()
    nthreads = mp.cpu_count()
    threads = [mp.Process(target=worker, args=(threadNo,))
        for threadNo in range(nthreads)]
    [t.start() for t in threads ]
    [t.join() for t in threads]
    